Bugfix for the konamiKeys attribute

xavisavvy opened the pull request:
While I was messing around with creating a wrapper for angular-konami.js I discovered that it was unable to interpret the objects being passed to it as an array, but was coming across as a string. 

Here is a simplified example:

**datDirective.js**

```
...
scope: {
    codeKeys: '='
 },
 template: '<hero konami-code="fireHadouken()" konami-keys="codeKeys"></hero>',
...
```

**view.html**

```
...
<dat-directive code-keys="[38,38]"></dat-directive>
...
```

With the above code, essentially datDirective picks up [38,38] as an array (verified using angular.isArray(scope.codeKeys). Then it stuffs it in the template and can also be seen as konami-keys="[38,38]" in the view. However when it gets to konamiCode.js it sees it as a string and breaks the array evaluation since it sees it as a string rather than an array object. 

Let me know if this doesn't make sense and I can get more detail. 

Overall a super simple implementation and I hope it makes its way onto npm ~ :+1:

Commit:
1. Fixed a bug when you are wanting to use this directive from another directive's template if you wanted to abstract it with a wrapper. Essentially angular-konami.js was not reading the value as an array but as a string which breaks the keypress evaluation.
2. Removed duplicate tag entry, and upped the version in the bower.json file

Commit:
Added WebStorm preference files to gitignore

lucasconstantino commented on angular-konami.js:
```diff
@@ -17,7 +20,7 @@ angular.module('konami', [])
 
         // Let user define a custom code.
-        var konamiKeys = attr.konamiKeys || konamiKeysDefault;
+        var konamiKeys = konamiKeys || konamiKeysDefault;
```
Hello, @xavisavvy! I don't really understand what have you modified here. It just seem to me that this code would break with an undefined exception. Shame on me not to have tests for this project, but isn't that right?

xavisavvy replied:
Hey!  no worries. Basically instead of using attr it is using an isolated scope (defined above in scope:{ konamiKeys:  '=' }) which also allows for javascript objects to be passed, as well as strings. I ran into this when I was using another directive that called this one. In its current state the following would work:

```
<div controller="controller as ctrl">
    <div konami-code="ctrl.foo()" konami-keys="[38,38]"></div>
</div>
```

However if I tried passing the array from the view's controller, it would not work since the konami-angular.js interprets it as a string rather than an array:

```
(in Controller)
ctrl.keys = [38,38];     
...
(In View)
<div controller="controller as ctrl">
    <div konami-code="ctrl.foo()" konami-keys="ctrl.keys"></div>
</div>
```

Making it an isolated/bound scope it should allow for dynamic swapping of the expected keys, and allows objects to be passed in. If they don't provide a konamiKey then it should just default to your konamiKeysDefault if I am not mistaken~ 

Hopefully that all made sense if not I can show you the working code example I am working with. As for tests I totally hear you! I rarely have put them in and I hope one day to be more disciplined and work with TDD :)

lucasconstantino commented:
That's an odd behavior. This must be something many people come across while using Angular. Could you try and find some reference to a similar problem in the Angular community?

Ps.: about the NPM, I can publish a new version promptly after merging the pull request. Thanks for the version bump, by the way. But in case you never had this need, know that you can have a NPM dependency pointing to a Git repository/branch, meaning you can already depend on your forked project while I don't publish the new version ;)

xavisavvy commented:
Good call on adding the git to NPM~ I will have to do that :) As for the problem in the Angular community I will poke around and see what I can find. I know some isolated scope variables can use @ to return strings exclusively but I was also confused by its behavior since I expected it to
bind to an object rather than a string.

UPDATE: I have been seeing some places where they say $attrs (for controller) and attrs (on link) for directives returns a string rather than an object which would in part explain things. I am looking further and will add a link here once i find a solid example. 

http://stackoverflow.com/questions/19079578/angular-directive-controller-attrs-value-as-object

xavisavvy commented:
Another update for those also searching the issue, most places I have been seeing in 'Official Documentation' the attrs object is shown but never actually used to assign objects. Most seem to utilize the isolated scope method but its unclear if that is due to the error we are seeing or if its just best practice. I'm looking through the open issues as well to see if I can attached this example to the core angular project as a question needing answers. 

https://docs.angularjs.org/guide/directive

xavisavvy commented:
Another update: I pulled a fresh AngularJS install so there is nothing but their skeleton app. I dropped angular-konami in with one small modification (I added the following lines to line 23 and 24):

```
console.log(angular.isArray(konamiKeys));
console.log(konamiKeys);
```

Outputs: 

```
angular-konami.js:23    false
angular-konami.js:24    [38, 38]
```

So I think attr.konamiKeys will always return a string rather than an array. Again not much documentation on how it returns, but rather people often use scope for anything object related. I guess it wasn't my directive wrapping it, its just how the attr works which breaks your code atm. 

I am working on further refining what I submitted so it will accept objects or a simple string. Hang tight~ 

Hope this helps! I can commit the testbed app if you want to pull it and have a look~

xavisavvy commented:
Been long enough that I think we both have lost track of the issue~ I will go ahead and close this out.
